add rule based auto tagging user guide (#9794)
opensearch-project/documentation-website: _tuning-your-cluster/availability-and-recovery/rule-based-autotagging/autotagging.md

Follow these best practices for designing and operating rule-based auto-tagging:

### Designing rules

When creating rules, focus on building logical, specific configurations that support your workload and access patterns. Consider the following guidelines:

* Identify the most relevant attributes for your use case.
* Use specific attribute values for precise control.
* Combine multiple attributes when appropriate.
* Use consistent naming conventions.
* Document attribute matching behavior.

### Managing attributes

Attribute selection and configuration significantly influence rule effectiveness. To manage attributes successfully, perform the following actions:

* Understand each attribute’s matching behavior.
* Start with the most specific criteria needed.
* Avoid overlapping rules unless intentional.
* Plan for future attribute value patterns.

### Operations

Ongoing operations and monitoring help maintain rule quality over time. Use the following practices to keep your setup reliable and effective:

* Test new rules in a development environment.
* Monitor rule matches in system logs.
* Document rule configurations.
* Regularly review rule effectiveness.
* Remove unused rules.

> ## Troubleshooting

Reviewer: @kaushalmahi12: We need a little more detail for each of these items to justify keeping this section. For example, when I need to "Verify attribute values and match rules" because no attributes have been assigned, how would I verify those attributes? The rules API?

Author: Sure, I will expand on this.

---
layout: default
title: Rule-based auto-tagging
nav_order: 20
parent: Workload management
grand_parent: Availability and recovery
---

# Rule-based auto-tagging

Rule-based auto-tagging automatically assigns feature-specific values to incoming requests and evaluates those requests against a set of predefined rules by matching request attributes.
For example, the workload management feature uses index patterns as an attribute and assigns workload group IDs.

Rule-based auto-tagging offers the following benefits:

* Flexible attribute-based matching
* Support for feature-specific matching logic
* Consistent policy application
* Automated request classification
* Reduced administrative overhead
* Centralized rule management
* Easy policy updates

Rule-based auto-tagging provides a flexible framework for implementing feature-specific request handling. Although this topic uses workload management as an example, the attribute-based matching system can be adapted for other OpenSearch features and use cases.
{: .tip }

## Key concepts

Before reviewing the rule configuration and behavior, it's important to understand the following key components of rule-based auto-tagging:

> * **Rule**: Defines matching criteria (attributes) and the value to assign

Reviewer 2: ```suggestion
* **Rule**: Defines matching criteria (attributes) and the value to assign.
```

> * **Attributes**: Key-value pairs used to match rules (such as index patterns, user roles, or request types)

Reviewer 2: ```suggestion
* **Attributes**: Key-value pairs used to match rules (such as index patterns, user roles, or request types).
```

> * **Feature-specific value**: The value assigned when a rule matches

Reviewer 2: ```suggestion
* **Feature-specific value**: The value assigned when a rule matches.
```

}
```

### Managing rules

Use the following API operations to manage rules for workload management:

#### Create or update a rule

```http
PUT /_rules/workload_group
{
  "index_patterns": ["prod-logs-*"],
  "other_attribute": ["value1"],
  "workload_group": "production_workload_id"
}
```

#### List rules

```http
GET /_rules/workload_group
```

#### Delete a rule

```http
DELETE /_rules/workload_group/{rule_id}
```

## Attribute matching

The attribute matching system determines which rules apply to a given request. Each attribute type can support different matching behaviors, based on the following attribute types:

1. **Exact matching**: Attribute values must match exactly
2. **Pattern matching**: Supports wildcards (such as index patterns)
3. **List matching**: Matches any item in a list
4. **Range matching**: Matches values within a defined range

For example, in workload management, index patterns support:

* Exact match: `logs-2025-04`
* Prefix pattern: `logs-2025-*`

Note: Matching behavior is determined by the feature and attribute type.

### Rule precedence

When multiple rules match a request, OpenSearch uses the following precedence rules:

> 1. Rules with more specific attribute matches take priority

Reviewer 2: ```suggestion
1. Rules with more specific attribute matches are prioritized.
```

> 2. Feature-specific tie-breaking logic is applied

Reviewer 2: ```suggestion
2. Feature-specific tie-breaking logic is applied.
```

For example, with index patterns:

> * `logs-prod-2025-*` takes precedence over `logs-prod-*`

Reviewer 2: ```suggestion
* `logs-prod-2025-*` takes precedence over `logs-prod-*`.
```

> * `logs-prod-*` takes precedence over `logs-*`

Reviewer 2: ```suggestion
* `logs-prod-*` takes precedence over `logs-*`.
```

### Evaluation process

OpenSearch evaluates incoming requests using the following process:

> 1. OpenSearch receives a request

Reviewer 2: ```suggestion
1. OpenSearch receives a request.
```

> 2. The system evaluates request attributes against defined rules

Reviewer 2: ```suggestion
2. The system evaluates request attributes against defined rules.
```

> 3. The most specific matching rule’s value is assigned

Reviewer 2: ```suggestion
3. The most specific matching rule's value is assigned.
```

> 4. If no rules match, no value is assigned

Reviewer 2: ```suggestion
4. If no rules match, no value is assigned.
```

## Examples

These examples demonstrate how rule-based auto-tagging works in workload management, which uses index patterns as its primary attribute.

### Multiple attribute matching

```json
{
  "index_patterns": ["logs-prod-*"],
  "request_type": ["search", "count"],
  "workload_group": "production_search_workload_id"
}

{
  "index_patterns": ["logs-prod-*"],
  "workload_group": "production_workload_id"
}
```

### Attribute specificity

```json
{
  "index_patterns": ["logs-*"],
  "workload_group": "general_workload_id"
}

{
  "index_patterns": ["logs-prod-service-*"],
  "workload_group": "prod_service_workload_id"
}
```

## Best practices

> Follow these best practices for designing and operating rule-based auto-tagging:

Reviewer 2: ```suggestion
Follow these best practices for designing and operating rule-based auto-tagging.
```

### Designing rules

When creating rules, focus on building logical, specific configurations that support your workload and access patterns. Consider the following guidelines:

* Identify the most relevant attributes for your use case.
* Use specific attribute values for precise control.
* Combine multiple attributes when appropriate.
* Use consistent naming conventions.
* Document attribute matching behavior.

### Managing attributes

Attribute selection and configuration significantly influence rule effectiveness. To manage attributes successfully, perform the following actions:

> * Understand each attribute’s matching behavior.

Reviewer 2: ```suggestion
* Understand each attribute's matching behavior.
```

* Start with the most specific criteria needed.
* Avoid overlapping rules unless intentional.
* Plan for future attribute value patterns.

### Operations

> Ongoing operations and monitoring help maintain rule quality over time. Use the following practices to keep your feature rules reliable and effective:

Reviewer 2: ```suggestion
Ongoing operations and monitoring help maintain rule quality over time. Use the following best practices to ensure that your feature rules are reliable and effective:
```
